### ISSUE TITLE
dylint-testing: Expose `CompileFail` mode from `compiletest`

### ISSUE THREAD

**Issue by jamie-osec** [trailofbits/dylint#1836]
This would allow making specific error annotations, making it easier to write tests that custom lints emit the expected errors and notes.

**Comment by smoelius** [comment]
You're referring to this? https://docs.rs/compiletest_rs/latest/compiletest_rs/common/enum.Mode.html#variant.CompileFail

I'm not familiar with it. Would you be able to give me a quick summary of how one would use it? Or could you point me to somewhere that it is used?

**Comment by jamie-osec** [comment]
Yes. It's useful to write, for example
```rs
#![deny(my_lint)]

let value = bad();  //~  NOTE `bad` created here
function(value());
//~^ ERROR `function` called with a value created in `bad`
```
And verify that the expected lints and notes are raised, and that other unexpected errors aren't raised (e.g. test against false-positives)

**Comment by smoelius** [comment]
Could you show me what your ideal API would look like?

**Comment by jamie-osec** [comment]
The most simple form would be an update to `ui::Test`, adding a new method to set the mode used when running the test.

**Comment by smoelius** [comment]
Switching to `ui_test` has been on my todo list for a while: #1111

And I think `ui_test` supports annotations like you describe in https://github.com/trailofbits/dylint/issues/1836#issuecomment-3552575169.

If `dylint_testing` were to wrap `ui_test` (instead of providing a new method), would that meet your needs?

If so, I will make switching a priority.

**Comment by jamie-osec** [comment]
Yes that would also be great
